### ISSUE TITLE
Consistent data-type output of DBS APIs in Go implementation

### ISSUE THREAD

**Issue by vkuznet** [dmwm/WMCore#10842]
**Impact of the new feature**
To perform migration to Go implementation of DBS server we need to verify that all DMWM/WMCore clients be able to parse `dbs2go` output which now will provide results in list data-type. The DBS Python implementation use dict, list and `null` for different types of output.

**Is your feature request related to a problem? Please describe.**
Migration to Go implementation of DBS server

**Describe the solution you'd like**
Ensure that DMWM/WMCore clients are able to parse `dbs2go` output

**Describe alternatives you've considered**
None

**Additional context**

DBS python server provides inconsistent data-type outputs, e.g. the exception is returned as dict, the results as list, and POST injection as `null`. In `dbs2go` implementation all results are returned as list data type. I suggest that we verify DMWM/WMCore clients to ensure  that they can properly parse provided API output. For example, the `dbs2go` error will be presented as following
```
https://cmsweb-testbed.cern.ch/dbs2go/filechildren
[{"api":"filechildren","error":"logical_file_name, block_id or block_name is required for fileparents api","exception":400,"method":"GET","type":"HTTPError"}]
```
while similar DBS python error looks like
```
https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader/filechildren
{"exception": 400, "message": "Logical_file_name, block_id or block_name is required for listFileChildren api", "type": "HTTPError"}
```
The difference is in list (`dbs2go`) vs dict (dbs pytthon)

**Comment by amaltaro** [comment]
@vkuznet Valentin, could you please clarify what is going to be the replacement for the `null` response of POST injection calls?

**Comment by vkuznet** [comment]
@amaltaro I suggest to standardize output from DBS server such that it should always return a list of dict objects. In case of empty results it returns empty list. So, I replaced `null` with empty list. This makes client code easier to maintain since you don't need to treat different outputs (like now DBS returns dict, list and null) and easily parse DBS server output.

**Comment by vkuznet** [comment]
another possibility is to return nothing at all since post injection request does not require output per-se if the request is successful. The client should use HTTP code instead, if it is not OK (the correct HTTP code should be 201 and not 200), then client should parse the output. But I found that DBS client code **always** parse the output instead of looking at HTTP code. It does that by calling cjson decode function. Therefore, to be complaint with DBS client I decided to return empty list.

**Comment by amaltaro** [comment]
Thanks for the clarification, Valentin! I agree with you, empty list is probably our best choice here.

**Comment by amaltaro** [comment]
Even though we haven't spotted any issues with the dbs server output, besides these two:
https://github.com/dmwm/WMCore/issues/10960
https://github.com/dmwm/WMCore/issues/10956

I think it would be safe to have a look at the following couple of modules:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/DBS/DBS3Reader.py
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/Tools/Common.py

to see whether any changes are required to deal with list-type output.

@vkuznet Valentin, if you want to check our code, please assign it to yourself and let us know what your findings are. Thanks

**Comment by vkuznet** [comment]
Yes, I'll check it.

**Comment by vkuznet** [comment]
@amaltaro , once I started testing DBS3Reader.py code I found that WMCore uses `listFileLumiArray` API which I was not aware of, and this API requires aggregation. I provided a fix to DBSClient, see https://github.com/dmwm/DBSClient/pull/66, and it means in order to proceed we'll need new version of DBSClient. I can quickly make a new DBSClient release, but I need to know if new release required for WMAgents, and if so how long we'll need (again) to wait for upgrade of WMAs? If it is not required and only part of WMCore, I can continue testing DBS3Reader APIs to see if I'll find any other hidden holes. Let me know.

**Comment by amaltaro** [comment]
@vkuznet thanks for spotting this issue.

After some investigation of our code, here is a summary:
* `listFileLumiArray` is embedded in our DBS3Reader.getParentFilesGivenParentDataset API; this API is also called from DBS3Reader.getParentFilesByLumi --> I do not see any code/service in WMCore actually using these 2 DBS3Reader APIs.
* `listFileLumiArray` is also embedded in DBS3Reader._getLumiList API; which is called from DBS3Reader.listFilesInBlock and DBS3Reader.listFilesInBlockWithParents --> these 2 APIs seem to be used by global and local workqueue(!)

I cannot understand why it didn't come up during the integration phase. It could be that I overlooked something, but I think the estimated number of lumis should have been quite off of the original/input values...

@vkuznet Valentin, if you have tested and you are confident with those changes in dbs3-client, please go ahead merging and cut a new release (4.0.6?). I need to look into the unit tests and better understand the differences and the impact of this.

Nonetheless, I think we will have to:
* patch dbs3-client in the WMAgent nodes
* release a new workqueue version and deploy it in CMSWEB testbed + production, depending on the new dbs3-client.

**Comment by amaltaro** [comment]
So, I investigated this issue with this unittest:
```
[dmwm@211f83ccdead DBS_t]$ nosetests DBSReader_t.py:DBSReaderTest.testListFilesInBlockWithParents
```

setting this unit test to be no longer emulated, thus defining the class like (and changing endpoint between production and testbed DBS Server):
```
class DBSReaderTest(unittest.TestCase):
```

DBS production - thus Python-based server - gives me an output like:
```
  'LogicalFileName': '/store/data/Commissioning2015/Cosmics/RECO/PromptReco-v1/000/238/545/00000/DA771768-95D0-E411-89C7-02163E0135C0.root',
  'LumiList': [{'LumiSectionNumber': [26,
                                      28,
                                      32,...
                                      42],
                'RunNumber': 238545}],
  'NumberOfEvents': 72123,...
  'ParentList': [{'LogicalFileName': '/store/data/Commissioning2015/Cosmics/RAW/v1/000/238/545/00000/1043E89F-2DCF-E411-9CAE-02163E013751.root',
                  'LumiList': [{'LumiSectionNumber': [30,
                                                      37,
                                                      39,...
                                                      41],
                                'RunNumber': 238545}],
                  'NumberOfEvents': 72123,
```

while testbed - thus Go-based server - gives me something like:
```
  'LogicalFileName': '/store/data/Commissioning2015/Cosmics/RECO/PromptReco-v1/000/238/545/00000/DA771768-95D0-E411-89C7-02163E0135C0.root',
  'LumiList': [{'EventCount': [None,
                               None,
                               None,...
                               None],
                'LumiSectionNumber': [26,
                                      28,
                                      32,...
                                      42],
                'RunNumber': 238545}],
  'NumberOfEvents': 72123,
  'ParentList': [{'LogicalFileName': '/store/data/Commissioning2015/Cosmics/RAW/v1/000/238/545/00000/1043E89F-2DCF-E411-9CAE-02163E013751.root',
                  'LumiList': [{'LumiSectionNumber': 30, 'RunNumber': 238545},
                               {'LumiSectionNumber': 37, 'RunNumber': 238545},
                               {'LumiSectionNumber': 39, 'RunNumber': 238545},...
                               {'LumiSectionNumber': 41, 'RunNumber': 238545}],
                  'NumberOfEvents': 72123,
```

and if I apply your DBSClient patch: https://github.com/dmwm/DBSClient/pull/66 , I start getting the same output as in the production server (python-based), with exception to the `EventCount` list of None's:
```
  'LogicalFileName': '/store/data/Commissioning2015/Cosmics/RECO/PromptReco-v1/000/238/545/00000/DA771768-95D0-E411-89C7-02163E0135C0.root',
  'LumiList': [{'EventCount': [None,
                               None,
                               None,
                               None],
                'LumiSectionNumber': [26,
                                      28,
                                      32,...
                                      42],
                'RunNumber': 238545}],
  'NumberOfEvents': 72123,
  'ParentList': [{'LogicalFileName': '/store/data/Commissioning2015/Cosmics/RAW/v1/000/238/545/00000/1043E89F-2DCF-E411-9CAE-02163E013751.root',
                  'LumiList': [{'EventCount': [None,
                                               None,
                                               None,
                                               None],
                                'LumiSectionNumber': [30,
                                                      37,
                                                      39,...
                                                      41],
                                'RunNumber': 238545}],
                  'NumberOfEvents': 72123,
```

In short, with the new DBSClient patch, data comes back in the same format, except for the `EventCount` list that is now present in the Go-based server(?)

**Comment by vkuznet** [comment]
@amaltaro , thanks for figure out the issue with test. The `EventCount` was added to server SQL query to simplify access to event count info for DAS (which is required by Rucio). Since it is an additional information it does not affect anything and safely ignore by machine tools.

Since I found at least one issue with DBS3Reader it would be more wise to let me finish testing this code (there are 42 APIs and different set of parameters to test), and then we can cut a new release of DBClient. Otherwise we may discover an additional stuff to fix and will need to redo whole things again.

**Comment by vkuznet** [comment]
@amaltaro , I finished testing and would like to report my findings:
- we identified that `listFileLumiArray` required changes in DBSClient due to missing aggregation. It is pity that DBSClient introduced different names for DBS server API. On the server nothing has changed since both `listFileLumis` and `listFileLumiArray` talked to `/filelumis` DBS server API
- I found that `crossCheck` API in DBS3Reader has a bug (it fails with Python server too) over [here](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/DBS/DBS3Reader.py#L343). The result `allLfns` is a list of dictionary with records containing `logical_file_name`. Therefore I fixed locally this code to be `setOfAllLfns = set([i['logical_file_name'] for i in allLfns])` and then the API tests were fine. I consider this as different issue which has nothing to do with Go  implementation. I can provide PR for it separately if you consider it relevant. My guess that no one really uses this API. I will leave decision up to you.
- I discovered few use-cases which were not covered neither by DBSClient unit tests or never reported by DMWM team during validation tests. All of them I fixed on a DBS Go server side while using DBS3Reader API calls. Here are the full list:
   - add support for block list in `/blockparents` API via POST request
   - fix SQL query for list of lfns in `/files` API (the issue was that DBS3Reader used `/files` API with dataset, run, lumi list parameters and their order in SQL query was wrong)
   - fix list `/datasets` API when querying with either LFN or run number (there was missing JOIN in SQL query)

I opened few dbs2go issues to cover these use-cases, see https://github.com/dmwm/dbs2go/issues/6, https://github.com/dmwm/dbs2go/issues/7 and https://github.com/dmwm/dbs2go/issues/8. All of them require additional unit tests which we'll provide.

Therefore, I'm glad we tested DBS3Reader as it uncover one aggregation issue and three additional use-cases which I was not aware.

Finally,  I create DBClient 4.0.6 release and uploaded it to PyPI which you may deploy to WMAs and other places.

**Comment by amaltaro** [comment]
Valentin, that's a pretty good outcome of this investigation.
Could you please create a new GH issue + PR for the `crossCheck` API?

> Finally, I create DBClient 4.0.6 release and uploaded it to PyPI which you may deploy to WMAs and other places.

Thanks! I'm about to define it under cmsdist as well and create a new WMCore patch release for central services and wmagent.

**Comment by amaltaro** [comment]
FYI.: I am working on this issue: https://github.com/dmwm/WMCore/issues/10983

**Comment by amaltaro** [comment]
@vkuznet Valentin, I have the `crossCheck` fix already in my local repository, together with many other unit tests that I have implemented. So please do not spend time on this.

**Comment by amaltaro** [comment]
@vkuznet I think we can consider this issue as resolved by now, after all the investigation and fixes in multiple places - as reported above - have been performed. Please reopen it if you see anything still missing to be tackled here. Thanks!
